README.md
# Masterthesis
This repository contains my masterthesis, the accompanying presentation, the evaluation scripts and raw data.

The thesis itself is [here](thesis/thesis.pdf), the presentation is [here](presentation/presentation.pdf).

The code is in my forks on GitHub:

- [llpc](https://github.com/Flakebi/llpc)
- [xgl](https://github.com/Flakebi/xgl)
- [llvm](https://github.com/Flakebi/amdvlk-llvm)
- [pal](https://github.com/Flakebi/pal)


CHANGELOG.md
# 2019-09-09 – 2019-09-15
- Many corrections and improvements
- Add address uniformity
- Add occupancy diagram
- Correct many, many small mistakes in the text
- Try to remove code from the infiltrator demo but it is not working

# 2019-09-02 – 2019-09-08
- Benchmark the switch-case vm
- Finish first text version of evaluation
- Add text in evaluation
- Draw diagrams for register usage
- Add diagrams about performance, finish overhead diagrams

# 2019-08-26 – 2019-09-01
- Change single lane executions to more robust method from atomic optimizations
- Add diagrams for unused code
- Automate data collection and diagram creation
- Add stubs for diagrams
- Finish text up to implementation for now

# 2019-08-19 – 2019-08-25
- Write text
- Generate table for overhead

# 2019-08-12 – 2019-08-18
- Add discussion about per lane vs per unit
- Add diagram for overhead
- Add examples for possible optimizations in design
- Disable shader cache for Feral games
- Try to run game benchmarks
	- **Ashes of the Singularity: Escalation**: Working
	- **Dota 2**: Working
	- **F1 2017**: Running, uses a shader cache and I don’t know where it is
	- F1 2018: Not starting
	- **Mad Max**: Running, uses a shader cache and I don’t know where it is
	- Serious Sam Fusion 2017: Running, but I cannot start the benchmarks automatically and inject environment variables
	- Total War Saga: Thrones of Britannia: Steam installs an empty folder
	- **Warhammer 40,000: Dawn of War III**: Running, uses a shader cache and I don’t know where it is
- Able to run Mad Max benchmark with the beta (https://www.reddit.com/r/linux_gaming/comments/8b6wyv/mad_max_benchmark_how/), run with `steam -applaunch 234140 --feral-banchmark`
- Unreal infiltrator works with PGO, uses more registers with PGO, little more than PGO when removing code
- Try to automatically start F1 2017
- Unreal infiltrator crashes when using PGO instrumentation
- Improve switch-vm, yields a 21% performance boost with unused-code-removal now
- Get non-uniformity count
- Dota has no non-uniform values for uniform addresses anymore

### 2019-07-29 – 2019-08-11

# 2019-07-22 – 2019-07-28
- Analyze uniformity of dota
- Can analyze addresses and loaded values for uniformity, early analysis is not perfectly accurate (addresses are marked as non-uniform early, but emit scalar load; switch not yet lowered to branches), works with late analysis (apart from loops)
- Use FunctionPass instead of ModulePass makes TargetTransformInfoWrapperPass available → LegacyDivergenceAnalysis works early

# 2019-07-15 – 2019-07-21
- Dead code removal improved performance of switch-vm by 0.1%
- Uniformity analyses works with late, per-lane
- Try to analyze switch-vm with RGP but there is not much to see
- Still compiling infiltrator demo, runs on linux

# 2019-07-08 – 2019-07-14
- Analyze uniformity of branches in dota and ashes: Everything is uniform…
- Start compiling infiltrator demo
- Count used registers when removing code for dota and switch-vm → tiny change
- Test removing dead code with switch-vm removes all branches but has no speed improvement
- Work on thesis: Improve diagrams, fix various todos

# 2019-07-01 – 2019-07-07
- Start installing UE4 and the infiltrator demo on windows
- Removing unused code works with dota and ashes, no changes in performance
- Write some text in the design section
- Removing unused code works with switch-case vm, not with dota
- Add citations to related work
- Add pass to remove unused code
- Try to build switch-case vm to test PGO performance but it is lowered to branches in IR (not SDAG), which does not use probabilities

# 2019-06-24 – 2019-06-30
- Capturing uniformity works, counter is 0 if variable is uniform
- Continue writing thesis
- Count unreachable code
- Improve presentation

# 2019-06-17 – 2019-06-23
- Benchmark: Need 11:30 for 2×6 dota benchmarks, 30 minutes for 2×3 ashes benchmarks
- GCC update breaks compilation
- Prepare presentation

# 2019-06-10 – 2019-06-16
- Start preparing presentation
- Use bb-counters for new "function" to count uniforms
- Per wave, normal instrumentation is wrong because of optimizations after setting the exec register, late instrumentation looks fine
- Using wave, normal instrumentation changed 1 shader in Dota (and linearization), wave-late changed more
- init_exec intrinsic does not work for some reason
- Write a script to automate things
- Cleanup code and introduce runtime options
- Plan finding uniform values
- Try to use gpuvis, overrides LD_LIBRARY_PATH and other things

# 2019-06-03 – 2019-06-09
- Create PGO analysis pass to find 0-counts
- LTTng: Can trace things, but how can it trace frames?
- Install games
- gpuvis: Supports Tracing a few frames but not continuously
- Benchmark Ashes
- Ashes never destroys pipelines → implement dumping every 10 seconds
- Create script to parse Ashes of the Singularity output files
- More Dota benchmarks

# 2019-05-27 – 2019-06-02
- OCAT runs on Windows only
- Steam claims negative fps, Ashes writes result to /mnt/bigdata/Dateien/Programme/Steam/steamapps/compatdata/507490/pfx/drive_c/users/steamuser/My Documents/My Games/Ashes of the Singularity - Escalation
- Start Ashes of the Singularity with proton from command line
- vkmark: No difference in pipelines
- Ask about VComputeBench
- 5 runs Dota benchmark: 11:54, everywhere 51.1 fps (89.7 for lower graphic settings)
- Increment counters only once per wave, increases fps while instrumenting to 35 in benchmark
- Scripted Dota benchmark, needs 6:48, 96 fps without PGO, 97.5 with PGO
- Problems
	- RGP always claims things are CPU limited, but it makes things run slower itself
	- Dota runs slow until reboot (again)
	- Dota demos have less fps (96) than real game (120)
- Run RGP, renderdoc, their integration is not working
- More Dota benchmarks, also with custom scene: 98 fps without PGO, 95 with PGO

# 2019-05-20 – 2019-05-26
- Benchmark Dota, no change in performance with PGO on the phoronix example

scripts/tex.py
#!/usr/bin/env python3
import argparse
from os import path

from data import *
from utils import *

games = {"dota": "Dota 2", "ashes": "Ashes", "dow3": "Warhammer",
		"f12017": "F1 2017", "madmax": "Mad Max", "switch": "Switch vm",
		"infiltrator": "Unreal Infiltrator"}

def overhead(args):
	configs = ["", "gen-wave-late", "gen-wave-late-non_atomic"]
	legend = ["Normal", "Atomic, per unit", "Non-atomic, per unit"]
	if args.diagram:
		overhead_diag(args, configs, legend)
	else:
		overhead_tab(args, configs, legend)

def overhead_diag(args, configs, legend):
	for config in configs:
		print(f"\\addplot+[error bars/.cd,y dir=both,y explicit] coordinates {{")
		for game in sorted(games.keys()):
			key = f"{game}-{config}"
			if key in bench and game != "switch":
				val = Value.avg([i * 1000 for i in bench[key]])
				print(f"({game},{val.val}) +- (0,{val.error})")

		print("};")

	print(f"\\legend{{{{{'},{'.join(legend)}}}}}")

def overhead_tab(args, configs, legend):
	print(r"""\begin{tabular}{l|l|l|l}
\textbf{Game} & \textbf{Config} & \textbf{Time per frame} & \textbf{Overhead}\\ """)

	for game in sorted(games.keys()):
		first = True
		for config, leg in zip(configs, legend):
			key = f"{game}-{config}"
			if key in bench and game != "switch":
				avg_val = Value.avg(bench[key])

				if first:
					overhead = ""
					start = f"\\hline\n{games[game]}"
					base = avg_val
					first = False
				else:
					overhead = ((avg_val / base - 1) * 100).tex_str("\\percent")
					start = ""

				avg_str = (avg_val * 1000).tex_str("\\milli\\second")
				print(f"""{start} & {leg} & {avg_str} & {overhead}\\\\\
""")

	print(r"\end{tabular}")

def counter_dist1(ctrs):
	bin_size = int(11222915000*2.5)
	bins = (max(ctrs) + bin_size - 1) // bin_size
	bins = 40
	bin_size = int(ceil(max(ctrs) / bins))

	bin_vals = [0] * bins
	for c in ctrs:
		if c != 0:
			bin_vals[int(c / bin_size)] += 1

	return bin_size, bin_vals

def counter_dist2(ctrs):
	bin_size = 0.0000016
	bins = int((10e4 / 1080 / 1920 / 999 + bin_size) / bin_size)

	bin_vals = [0] * bins
	for c in ctrs:
		if c != 0:
			i = int(c / bin_size)
			if i < len(bin_vals):
				bin_vals[i] += 1

	return bin_size, bin_vals

def counter_dist(args, f):
	if not args.game:
		print("Need to get a game for counter distribution")
		return

	for k, v in counters.items():
		if not k.startswith(args.game):
			continue

		# Divide through screen resolution and amount of frames
		if args.game != "dota":
			raise Exception("Only dota is supported, please insert the number of frames")

		ctrs = [item / 1080 / 1920 / 999 for l in v for item in l]

		print(f"""
\\addplot[
	fill=tumblue,
] table[header=false, row sep=\\\\] {{\
""")

		bin_size, bin_vals = f(ctrs)

		for i, b in enumerate(bin_vals):
			print(f"{i*bin_size + bin_size/2}\t{b}\\\\")

		print("} \closedcycle;")

def get_non_zero(v):
	v2 = []
	for pipeline in v:
		zeroes = 0
		for c in pipeline:
			if c == 0:
				zeroes += 1

		if zeroes != len(pipeline):
			v2.append(pipeline)
	return v2

def unused_code(args):
	if not args.game:
		print("Need to get a game for unused code")
		return

	for k, v in counters.items():
		if not k.startswith(args.game):
			continue

		print(f"""
\\addplot[
	fill=tumblue,
] table[header=false, row sep=\\\\] {{\
""")

		v2 = get_non_zero(v)

		bins = max([len(l) for l in v2]) + 1
		bin_size = 1
		bin_zero = [0] * bins
		bin_bbs = [0] * bins

		for pipeline in v2:
			i = len(pipeline) // bin_size
			zeroes = 0
			for c in pipeline:
				if c == 0:
					zeroes += 1

			bin_zero[i] += zeroes
			bin_bbs[i] += len(pipeline)

		for i in range(bins):
			print(f"{i*bin_size + bin_size//2}\t{safe_divide(bin_zero[i], bin_bbs[i]) * 100}\\\\")

		print("} \closedcycle;")

def bbs(args):
	if not args.game:
		print("Need to get a game for unused code")
		return

	for k, v in counters.items():
		if not k.startswith(args.game):
			continue

		print(f"""
\\addplot[
	smooth,
	fill=gray,
	fill opacity=0.4,
	draw=none,
] table[header=false, row sep=\\\\] {{\
""")

		v2 = get_non_zero(v)

		bins = 50
		bin_size = max([len(l) for l in v2]) // bins + 1

		bin_vals = [0] * bins

		for pipeline in v2:
			i = len(pipeline) // bin_size
			bin_vals[i] += 1

		for i in range(bins - 1, -1, -1):
			if bin_vals[i] != 0:
				break
			bin_vals.pop()

		for i in range(len(bin_vals)):
			print(f"{i*bin_size + bin_size//2}\t{bin_vals[i] / len(v2) / bin_size * 100}\\\\")

		print("} \closedcycle;")

def max_bbs(args):
	if not args.game:
		print("Need to get a game for max bbs")
		return

	for k, v in counters.items():
		if not k.startswith(args.game):
			continue

		v2 = get_non_zero(v)
		bins = max([len(l) for l in v2]) + 1
		print(f"{bins}")

def unused_code_summary(args):
	used_games = []
	for game in sorted(games.keys()):
		for k, v in counters.items():
			if not k.startswith(game):
				continue

			used_games.append((game, v))

	print(f"""
\\begin{{axis}}[
	ybar,
	enlarge x limits=0.1,
	symbolic x coords={{{",".join([g for g, _ in used_games])}}},
	xticklabels={{{",".join([games[g] for g, _ in used_games])}}},
	xtick=data,
	xticklabel style={{text width=1.5cm,align=center}},
	ylabel={{Unused blocks [\SI{{}}{{\percent}}]}},
	ymin=0,
	ymax=119,
	grid=both,
	nodes near coords,
	every node near coord/.append style={{rotate=90, anchor=west}},
	bar width=0.5cm,
	legend style={{at={{(1,1)}},anchor=north east}},
]

\\addplot[
	fill=tumblue,
] table[header=false, row sep=\\\\] {{\
""")

	for game, v in used_games:
		v2 = get_non_zero(v)

		all_bbs = 0
		zero_bbs = 0
		for pipeline in v2:
			all_bbs += len(pipeline)
			for c in pipeline:
				if c == 0:
					zero_bbs += 1

		print(f"{game}\t{zero_bbs / all_bbs * 100}\\\\")

	print("};\n\\end{axis}")

def performance(args):
	configs = ["", "use-wave-late", "use-wave-late-remove"]
	legend = ["Normal", "PGO", "PGO + removing blocks"]
	if args.diagram:
		performance_diag(args, configs, legend)
	else:
		performance_tab(args, configs, legend)

def performance_diag(args, configs, legend):
	for config in configs:
		print(f"\\addplot+[error bars/.cd,y dir=both,y explicit] coordinates {{")
		for game in sorted(games.keys()):
			key = f"{game}-{config}"
			if key in bench:
				val = Value.avg([i * 1000 for i in bench[key]])
				print(f"({game},{val.val}) +- (0,{val.error})")

		print("};")

	print(f"\\legend{{{{{'},{'.join(legend)}}}}}")

def performance_tab(args, configs, legend):
	print(r"""\begin{tabular}{l|l|l|l}
\textbf{Game} & \textbf{Config} & \textbf{Time per frame} & \textbf{Difference}\\ """)

	for game in sorted(games.keys()):
		first = True
		for config, leg in zip(configs, legend):
			key = f"{game}-{config}"
			if key in bench:
				avg_val = Value.avg(bench[key])

				if first:
					overhead = ""
					start = f"\\hline\n{games[game]}"
					base = avg_val
					first = False
				else:
					overhead = ((avg_val / base - 1) * 100).tex_str("\\percent")
					start = ""

				avg_str = (avg_val * 1000).tex_str("\\milli\\second")
				print(f"""{start} & {leg} & {avg_str} & {overhead}\\\\\
""")

	print(r"\end{tabular}")

def registers_fun(args):
	configs = ["", "use-wave-late", "use-wave-late-remove"]
	legend = ["Normal", "PGO", "PGO + removing blocks"]

	if not args.game:
		print("Need to get a type (cs/vs/ps) for registers")
		return

	typ, reg_typ = args.game.split("-")
	if typ == "cs":
		start_index = 0
	elif typ == "vs":
		start_index = 2
	elif typ == "ps":
		start_index = 4
	else:
		print(f"{typ} is a wrong type, try cs/vs/ps")
		return

	if reg_typ == "sgpr":
		start_index += 1
	elif reg_typ != "vgpr":
		print(f"{reg_typ} is a wrong type, try vgpr/sgpr")
		return

	# Only compare pipelines which occur in all configurations
	pipelines = {}
	for game in sorted(games.keys()):
		pipelines[game] = {}
		config_count = 0
		for config in configs:
			key = f"{game}-{config}"
			if key in registers:
				config_count += 1
				regs = registers[key]
				for pipe, v in regs.items():
					if v[start_index] != 0:
						if pipe not in pipelines[game]:
							pipelines[game][pipe] = 1
						else:
							pipelines[game][pipe] += 1
		# Remove pipelines which are only available in some configs
		ps = [p for p, v in pipelines[game].items() if v != config_count]
		for p in ps:
			del pipelines[game][p]

	for config in configs:
		print("\\addplot coordinates {")
		for game in sorted(games.keys()):
			key = f"{game}-{config}"
			if key in registers:
				regs = registers[key]
				count = 0
				amount = 0
				for p in pipelines[game].keys():
					v = regs[p]
				# for v in regs.values():
					if v[start_index] != 0:
						count += v[start_index]
						amount += 1
				if amount != 0:
					count /= amount

				print(f"({game},{count})")
		print("};")

	print(f"\\legend{{{{{'},{'.join(legend)}}}}}")

def uniform(args, key):
	configs = ["static", "dynamic", "divergent"]
	amount = {game: 0 for game in games}
	counters = {}
	for config in configs:
		for game in games.keys():
			counter = 0
			uni_key = f"{game}-use-wave-late-uniform"
			if uni_key not in uniformity:
				continue

			for shader in uniformity[uni_key]:
				counter += shader[config][key]

			if not config in counters:
				counters[config] = {}
			counters[config][game] = counter
			amount[game] += counter

	for config in configs:
		print("\\addplot coordinates {")
		for game in sorted(games.keys()):
			if game in counters[config]:
				count = counters[config][game] / amount[game] * 100
				print(f"({game},{count})")
		print("};")

def uniform_branches(args):
	uniform(args, "Condition")

def uniform_loads(args):
	uniform(args, "LoadValue")

def uniform_addrs(args):
	uniform(args, "Address")

def register_scatter(args):
	# Analyze vgprs only
	xconfig = "use-wave-late"
	yconfig = "use-wave-late-remove"
	indices = [0, 2, 4]
	games = ["ashes", "dota", "infiltrator", "madmax"]

	print(r"""
\addplot[scatter,only marks,scatter src=explicit symbolic]
coordinates {""")
	for game in games:
		xdata = registers[f"{game}-{xconfig}"]
		ydata = registers[f"{game}-{yconfig}"]
		for k in xdata.keys():
			if k not in ydata:
				continue

			# Count of shader types used in the pipeline
			for i in indices:
				if xdata[k][i] != 0:
					print(f"({xdata[k][i]},{ydata[k][i]}) [{game}]")

	print("};")

def get_occupancy(regs):
	max_regs = 256

	# Round registers up to a multiple of 4
	regs = (regs + 3) // 4 * 4
	return max_regs // regs

def register_occupancy(args):
	# Analyze vgprs only
	xconfig = "use-wave-late"
	yconfig = "use-wave-late-remove"
	indices = [0, 2, 4]
	games = ["ashes", "dota", "infiltrator", "madmax"]

	data_size = 10
	print(f"""
\\addplot [matrix plot*,point meta=explicit,mesh/cols={data_size},mesh/rows={data_size}]
table [meta index=2,header=false,row sep=\\\\] {{""")
	data = [[0 for _ in range(data_size)] for _ in range(data_size)]
	for game in games:
		xdata = registers[f"{game}-{xconfig}"]
		ydata = registers[f"{game}-{yconfig}"]
		for k in xdata.keys():
			if k not in ydata:
				continue

			# Count of shader types used in the pipeline
			for i in indices:
				if xdata[k][i] != 0:
					x = get_occupancy(xdata[k][i])
					y = get_occupancy(ydata[k][i])
					if y <= data_size and x <= data_size:
						data[y - 1][x - 1] += 1

	for y, row in enumerate(data):
		for x, val in enumerate(row):
			print(f"{x + 1}\t{y + 1}\t{val}\\\\")

	print("};")

def data(args):
	xconfig = "use-wave-late"
	indices = [0, 2, 4]

	for game in sorted(games.keys()):
		print(f"\n{game}")

		# Count shaders
		count = 0
		xdata = registers[f"{game}-{xconfig}"]
		for x in xdata.values():
			for i in indices:
				if x[i] != 0:
					count += 1
		print(f"{count} shaders")

def main():
	actions = {
		"overhead": overhead,
		"counter-dist1": lambda args: counter_dist(args, counter_dist1),
		"counter-dist2": lambda args: counter_dist(args, counter_dist2),
		"unused-code": unused_code,
		"bbs": bbs,
		"max-bbs": max_bbs,
		"unused-code-summary": unused_code_summary,
		"performance": performance,
		"registers": registers_fun,
		"uniform-branches": uniform_branches,
		"uniform-loads": uniform_loads,
		"uniform-addrs": uniform_addrs,
		"register-scatter": register_scatter,
		"register-occupancy": register_occupancy,
		"data": data,
	}

	parser = argparse.ArgumentParser(description="Generate tex code")
	parser.add_argument("action", choices=list(actions.keys()))
	parser.add_argument("-d", "--diagram", action="store_true", help="Emit diagram instead of table")
	parser.add_argument("-g", "--game")

	args = parser.parse_args()

	if args.action in actions:
		actions[args.action](args)

if __name__ == "__main__":
	main()
